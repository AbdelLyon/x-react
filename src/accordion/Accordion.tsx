import { forwardRef } from "react";
import {
  Accordion as AccordionRoot,
  AccordionItem,
  AccordionProps,
  AccordionItemProps,
} from "@nextui-org/react";
import { cn } from "@/utils";

interface ExtendedAccordionItemProps
  extends Omit<AccordionItemProps, "content"> {
  content?: React.ReactNode;
}

interface AccordionWrapperProps extends Omit<AccordionProps, "children"> {
  items: ExtendedAccordionItemProps[];
  itemClasses?: {
    base?: string;
    title?: string;
    [key: string]: string | undefined;
  };
}

export const Accordion = forwardRef<HTMLDivElement, AccordionWrapperProps>(
  ({ items, itemClasses, ...accordionProps }, ref) => {
    const defaultItemClasses = {
      base: "w-full border-none shadow-lg dark:shadow-none",
      title: "text-lg font-semibold",
    };

    return (
      <AccordionRoot
        ref={ref}
        {...accordionProps}
        itemClasses={{
          ...defaultItemClasses,
          ...itemClasses,
          base: cn(defaultItemClasses.base, itemClasses?.base),
          title: cn(defaultItemClasses.title, itemClasses?.title),
        }}
      >
        {items.map((item) => {
          const { content, ...itemProps } = item;
          return (
            <AccordionItem
              {...itemProps}
              key={itemProps.key}
              children={content}
            />
          );
        })}
      </AccordionRoot>
    );
  },
);

Accordion.displayName = "Accordion";
