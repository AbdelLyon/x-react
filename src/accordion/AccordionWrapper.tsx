import {
  Accordion,
  AccordionItem,
  AccordionProps,
  AccordionItemProps,
} from "@nextui-org/react";

interface AccordionWrapperProps extends AccordionProps {
  items: AccordionItemProps[];
}

export const AccordionWrapper = ({
  items,
  ...accordionProps
}: AccordionWrapperProps) => {
  return (
    <Accordion {...accordionProps}>
      {items.map((item) => {
        return (
          <AccordionItem {...item} key={item.key}>
            {item.content}
          </AccordionItem>
        );
      })}
    </Accordion>
  );
};
