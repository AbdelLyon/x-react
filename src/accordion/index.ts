import "@/index.css";
export { Accordion } from "@/accordion/Accordion";
