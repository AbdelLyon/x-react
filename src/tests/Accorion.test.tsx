import { Accordion } from "@/accordion";
import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";

describe("Accordion", () => {
  const defaultItems = [
    {
      key: "1",
      title: "First Item",
      content: "Content for first item",
    },
    {
      key: "2",
      title: "Second Item",
      content: "Content for second item",
    },
  ];

  it("renders all accordion items", () => {
    const { getAllByRole } = render(<Accordion items={defaultItems} />);
    const accordionButtons = getAllByRole("button");
    expect(accordionButtons).toHaveLength(defaultItems.length);
  });

  it("displays correct titles", () => {
    const { getByText } = render(<Accordion items={defaultItems} />);
    defaultItems.forEach((item) => {
      expect(getByText(item.title)).toBeTruthy();
    });
  });

  it("applies default classes", () => {
    const { container } = render(<Accordion items={defaultItems} />);
    const accordionItems = container.querySelectorAll('[class*="text-lg"]');
    expect(accordionItems.length).toBeGreaterThan(0);
  });

  it("applies custom item classes", () => {
    const { container } = render(
      <Accordion
        items={defaultItems}
        itemClasses={{
          base: "custom-base-class",
          title: "custom-title-class",
        }}
      />,
    );

    const baseClasses = container.querySelectorAll(".custom-base-class");
    const titleClasses = container.querySelectorAll(".custom-title-class");
    expect(baseClasses.length).toBeGreaterThan(0);
    expect(titleClasses.length).toBeGreaterThan(0);
  });

  it("handles splitted variant", () => {
    const { container } = render(
      <Accordion items={defaultItems} variant="splitted" />,
    );

    const splittedItems = container.querySelectorAll(
      '[class*="border-1"][class*="rounded-md"]',
    );
    expect(splittedItems.length).toBeGreaterThan(0);
  });

  it("renders with minimal props", () => {
    const minimalItems = [{ key: "1", title: "Minimal Item" }];
    const { getByText } = render(<Accordion items={minimalItems} />);
    expect(getByText("Minimal Item")).toBeTruthy();
  });

  it("handles empty items array", () => {
    const { container } = render(<Accordion items={[]} />);
    expect(container.firstChild?.childNodes.length).toBe(0);
  });

  it("passes through accordion props", () => {
    const { container } = render(
      <Accordion
        items={defaultItems}
        selectionMode="multiple"
        data-selection-mode="multiple"
      />,
    );

    const accordionRoot = container.firstChild;
    expect(accordionRoot).toHaveAttribute("data-selection-mode", "multiple");
  });
});
