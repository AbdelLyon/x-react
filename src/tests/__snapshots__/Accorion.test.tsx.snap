// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion > matches snapshot with default props 1`] = `
<DocumentFragment>
  <div
    class="px-2 w-full"
    data-orientation="vertical"
  >
    <div
      class="w-full shadow-none"
    >
      <h2>
        <button
          aria-expanded="false"
          class="flex py-4 w-full h-full gap-3 items-center tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 transition-opacity"
          id=":ri:"
          type="button"
        >
          <div
            class="flex-1 flex flex-col text-start"
          >
            <span
              class="text-foreground text-lg font-semibold"
            >
              First Item
            </span>
          </div>
          <span
            aria-hidden="true"
            class="text-default-400 transition-transform rotate-0 data-[open=true]:-rotate-90 rtl:-rotate-180 rtl:data-[open=true]:-rotate-90"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="1em"
              role="presentation"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M15.5 19l-7-7 7-7"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
            </svg>
          </span>
        </button>
      </h2>
    </div>
    <hr
      class="shrink-0 bg-divider border-none w-full h-divider"
      role="separator"
    />
    <div
      class="w-full shadow-none"
    >
      <h2>
        <button
          aria-expanded="false"
          class="flex py-4 w-full h-full gap-3 items-center tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 transition-opacity"
          id=":rk:"
          type="button"
        >
          <div
            class="flex-1 flex flex-col text-start"
          >
            <span
              class="text-foreground text-lg font-semibold"
            >
              Second Item
            </span>
          </div>
          <span
            aria-hidden="true"
            class="text-default-400 transition-transform rotate-0 data-[open=true]:-rotate-90 rtl:-rotate-180 rtl:data-[open=true]:-rotate-90"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="1em"
              role="presentation"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M15.5 19l-7-7 7-7"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
            </svg>
          </span>
        </button>
      </h2>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Accordion > matches snapshot with full configuration 1`] = `
<DocumentFragment>
  <div
    class="px-2 flex flex-col gap-2 w-full"
    data-orientation="vertical"
  >
    <div
      class="px-4 w-full shadow-none bg-white dark:bg-content1 border-1 border-default-200 rounded-md custom-base"
    >
      <h2>
        <button
          aria-expanded="false"
          class="flex py-4 w-full h-full gap-3 items-center tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 transition-opacity"
          id=":rm:"
          type="button"
        >
          <div
            class="flex-1 flex flex-col text-start"
          >
            <span
              class="text-foreground text-lg font-semibold custom-title"
            >
              First Item
            </span>
          </div>
          <span
            aria-hidden="true"
            class="text-default-400 transition-transform rotate-0 data-[open=true]:-rotate-90 rtl:-rotate-180 rtl:data-[open=true]:-rotate-90"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="1em"
              role="presentation"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M15.5 19l-7-7 7-7"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
            </svg>
          </span>
        </button>
      </h2>
    </div>
    <div
      class="px-4 w-full shadow-none bg-white dark:bg-content1 border-1 border-default-200 rounded-md custom-base"
    >
      <h2>
        <button
          aria-expanded="false"
          class="flex py-4 w-full h-full gap-3 items-center tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 transition-opacity"
          id=":ro:"
          type="button"
        >
          <div
            class="flex-1 flex flex-col text-start"
          >
            <span
              class="text-foreground text-lg font-semibold custom-title"
            >
              Second Item
            </span>
          </div>
          <span
            aria-hidden="true"
            class="text-default-400 transition-transform rotate-0 data-[open=true]:-rotate-90 rtl:-rotate-180 rtl:data-[open=true]:-rotate-90"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="1em"
              role="presentation"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M15.5 19l-7-7 7-7"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
            </svg>
          </span>
        </button>
      </h2>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should match snapshot with custom fallback 1`] = `
<DocumentFragment>
  <div
    class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full"
    tabindex="-1"
  >
    <div
      aria-label="avatar"
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex items-center justify-center"
      role="img"
    >
      <div>
        Custom Fallback
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should match snapshot with default props 1`] = `
<DocumentFragment>
  <div
    class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full"
    tabindex="-1"
  >
    <span
      aria-label="avatar"
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex items-center justify-center text-inherit w-full h-full"
      role="img"
    >
      <svg
        aria-hidden="true"
        fill="none"
        height="80%"
        role="presentation"
        viewBox="0 0 24 24"
        width="80%"
      >
        <path
          d="M12 2C9.38 2 7.25 4.13 7.25 6.75C7.25 9.32 9.26 11.4 11.88 11.49C11.96 11.48 12.04 11.48 12.1 11.49C12.12 11.49 12.13 11.49 12.15 11.49C12.16 11.49 12.16 11.49 12.17 11.49C14.73 11.4 16.74 9.32 16.75 6.75C16.75 4.13 14.62 2 12 2Z"
          fill="currentColor"
        />
        <path
          d="M17.0809 14.1489C14.2909 12.2889 9.74094 12.2889 6.93094 14.1489C5.66094 14.9989 4.96094 16.1489 4.96094 17.3789C4.96094 18.6089 5.66094 19.7489 6.92094 20.5889C8.32094 21.5289 10.1609 21.9989 12.0009 21.9989C13.8409 21.9989 15.6809 21.5289 17.0809 20.5889C18.3409 19.7389 19.0409 18.5989 19.0409 17.3589C19.0309 16.1289 18.3409 14.9889 17.0809 14.1489Z"
          fill="currentColor"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should match snapshot with image 1`] = `
<DocumentFragment>
  <div
    class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full"
    tabindex="-1"
  >
    <img
      alt="Test Avatar"
      class="flex object-cover w-full h-full transition-opacity !duration-500 opacity-0 data-[loaded=true]:opacity-100"
      src="https://example.com/avatar.jpg"
    />
  </div>
</DocumentFragment>
`;

exports[`Avatar > should match snapshot with name 1`] = `
<DocumentFragment>
  <div
    class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full"
    tabindex="-1"
  >
    <span
      aria-label="John Doe"
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 font-normal text-center text-inherit"
      role="img"
    >
      Joh
    </span>
  </div>
</DocumentFragment>
`;
